### ISSUE TITLE
Context manager not raising exception but simply returns False

### ISSUE THREAD

**Issue by bholmgren** [napalm-automation/napalm-base#297]
In the commit dbddbbbbe422ebbe66e995f06518d2a701a5cb3c the context manager was broken and no longer raises an exception in case of failure opening a connection to the device. Instead it returns False which effectively means I have to check for False and make up my own exception. This is not very useful.

```
In [1]: from napalm_base import get_network_driver
In [2]: eos=get_network_driver('eos')
In [3]: with eos('mydevice, 'user', 'pass') as device:
   ...:     print device
   ...:
False
```

I would expect this to raise the exception 

> ConnectionException: unable to connect to eAPI

**Comment by bewing** [comment]
Behavior at e72c522c60530fcedd4dc47b66d7a566c828852f:
```python
In [1]: from napalm_eos import EOSDriver

In [2]: with EOSDriver('a', 'a', 'a') as device:
   ...:     pass
   ...:
[Errno -2] Name or service not known
Traceback (most recent call last):
  File "/home/bewing/.virtualenvs/napalm-base/lib/python3.6/site-packages/pyeapi/eapilib.py", line 385, in send
    self.transport.endheaders(message_body=data)
  File "/usr/local/lib/python3.6/http/client.py", line 1234, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "/usr/local/lib/python3.6/http/client.py", line 1026, in _send_output
    self.send(msg)
  File "/usr/local/lib/python3.6/http/client.py", line 964, in send
    self.connect()
  File "/usr/local/lib/python3.6/http/client.py", line 1392, in connect
    super().connect()
  File "/usr/local/lib/python3.6/http/client.py", line 936, in connect
    (self.host,self.port), self.timeout, self.source_address)
  File "/usr/local/lib/python3.6/socket.py", line 704, in create_connection
    for res in getaddrinfo(host, port, 0, SOCK_STREAM):
  File "/usr/local/lib/python3.6/socket.py", line 743, in getaddrinfo
    for res in _socket.getaddrinfo(host, port, family, type, proto, flags):
socket.gaierror: [Errno -2] Name or service not known
---------------------------------------------------------------------------
gaierror                                  Traceback (most recent call last)
~/.virtualenvs/napalm-base/lib/python3.6/site-packages/pyeapi/eapilib.py in send(self, data)
    384
--> 385             self.transport.endheaders(message_body=data)
    386

/usr/local/lib/python3.6/http/client.py in endheaders(self, message_body, encode_chunked)
   1233             raise CannotSendHeader()
-> 1234         self._send_output(message_body, encode_chunked=encode_chunked)
   1235

/usr/local/lib/python3.6/http/client.py in _send_output(self, message_body, encode_chunked)
   1025         del self._buffer[:]
-> 1026         self.send(msg)
   1027

/usr/local/lib/python3.6/http/client.py in send(self, data)
    963             if self.auto_open:
--> 964                 self.connect()
    965             else:

/usr/local/lib/python3.6/http/client.py in connect(self)
   1391
-> 1392             super().connect()
   1393

/usr/local/lib/python3.6/http/client.py in connect(self)
    935         self.sock = self._create_connection(
--> 936             (self.host,self.port), self.timeout, self.source_address)
    937         self.sock.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)

/usr/local/lib/python3.6/socket.py in create_connection(address, timeout, source_address)
    703     err = None
--> 704     for res in getaddrinfo(host, port, 0, SOCK_STREAM):
    705         af, socktype, proto, canonname, sa = res

/usr/local/lib/python3.6/socket.py in getaddrinfo(host, port, family, type, proto, flags)
    742     addrlist = []
--> 743     for res in _socket.getaddrinfo(host, port, family, type, proto, flags):
    744         af, socktype, proto, canonname, sa = res

gaierror: [Errno -2] Name or service not known

During handling of the above exception, another exception occurred:

ConnectionError                           Traceback (most recent call last)
~/.virtualenvs/napalm-base/lib/python3.6/site-packages/napalm_eos/eos.py in open(self)
    112             # let's try to run a very simple command
--> 113             self.device.run_commands(['show clock'], encoding='text')
    114         except ConnectionError as ce:

~/.virtualenvs/napalm-base/lib/python3.6/site-packages/pyeapi/client.py in run_commands(self, commands, encoding, send_enable, **kwargs)
    729
--> 730         response = self._connection.execute(commands, encoding, **kwargs)
    731

~/.virtualenvs/napalm-base/lib/python3.6/site-packages/pyeapi/eapilib.py in execute(self, commands, encoding, **kwargs)
    494             request = self.request(commands, encoding=encoding, **kwargs)
--> 495             response = self.send(request)
    496             return response

~/.virtualenvs/napalm-base/lib/python3.6/site-packages/pyeapi/eapilib.py in send(self, data)
    427                              % str(exc))
--> 428             raise ConnectionError(str(self), error_msg)
    429

ConnectionError: Socket error during eAPI connection: [Errno -2] Name or service not known

During handling of the above exception, another exception occurred:

ConnectionException                       Traceback (most recent call last)
<ipython-input-9-255fc80f5618> in <module>()
----> 1 with EOSDriver('a', 'a', 'a') as device:
      2     pass

/mnt/c/Users/bewing/PycharmProjects/napalm-base/napalm_base/base.py in __enter__(self)
     52         except:  # noqa
     53             exc_info = sys.exc_info()
---> 54             self.__raise_clean_exception(exc_info[0], exc_info[1], exc_info[2])
     55         return self
     56

/mnt/c/Users/bewing/PycharmProjects/napalm-base/napalm_base/base.py in __raise_clean_exception(exc_type, exc_value, exc_traceback)
     92             print(epilog)
     93         # Traceback should already be attached to exception; no need to re-attach
---> 94         raise exc_value
     95
     96     def open(self):

/mnt/c/Users/bewing/PycharmProjects/napalm-base/napalm_base/base.py in __enter__(self)
     49     def __enter__(self):
     50         try:
---> 51             self.open()
     52         except:  # noqa
     53             exc_info = sys.exc_info()

~/.virtualenvs/napalm-base/lib/python3.6/site-packages/napalm_eos/eos.py in open(self)
    116             # either if HTTP(S) agent is not enabled
    117             # show management api http-commands
--> 118             raise ConnectionException(ce.message)
    119
    120     def close(self):

ConnectionException: Socket error during eAPI connection: [Errno -2] Name or service not known
```

**Comment by bewing** [comment]
This appears to be an issue only in python2.7:
```python
Python 2.7.12 (default, Nov 19 2016, 06:48:10)
Type "copyright", "credits" or "license" for more information.

IPython 5.4.1 -- An enhanced Interactive Python.
?         -> Introduction and overview of IPython's features.
%quickref -> Quick reference.
help      -> Python's own help system.
object?   -> Details about 'object', use 'object??' for extra details.

In [1]: from napalm_eos import EOSDriver

In [2]: with EOSDriver('www.google.com', 'a', 'a') as device:
   ...:     pass
   ...:
No handlers could be found for logger "pyeapi.eapilib"

In [3]: exit
```
```python
Python 3.6.2 (default, Aug  7 2017, 14:15:33)
Type 'copyright', 'credits' or 'license' for more information
IPython 6.1.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: from napalm_eos import EOSDriver

In [2]: with EOSDriver('www.google.com', 'a', 'a') as device:
   ...:     pass
   ...:
Expecting value: line 1 column 1 (char 0)
Traceback (most recent call last):
  File "/home/bewing/.virtualenvs/test/lib/python3.6/site-packages/pyeapi/eapilib.py", line 402, in send
    decoded = json.loads(response_content)
  File "/usr/local/lib/python3.6/json/__init__.py", line 354, in loads
    return _default_decoder.decode(s)
  File "/usr/local/lib/python3.6/json/decoder.py", line 339, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/local/lib/python3.6/json/decoder.py", line 357, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
```

**Comment by dbarrosop** [comment]
The context manager has to return `False` when an exception occurred and it wasn't recovered:

http://effbot.org/zone/python-with-statement.htm

There might be some other issue though.

**Comment by dbarrosop** [comment]
Ok, you were right, looks like that commit returned False in `__enter__` by mistake. `False` has to be returned only on `__exit__`.

**Comment by dbarrosop** [comment]
Could you test #299, please?

**Comment by dbarrosop** [comment]
To make it more interesting looks like previous implementation was actually good for `py3`. I guess that's where the confusion originally came from... So standby until we get this fixed correctly for both `py2` and `py3`.

**Comment by bholmgren** [comment]
#299 does work now in py2.


Perhaps avoiding directly returning the value from __raise_clean_exception() and be explicit with the return values in __enter__() and __exit__() is advisable. Returning values from functions with callers of different needs can be risky as a change to satisfy one of the callers may have an impact on the other.
I guess what I mean is I'd suggest not returning anything from __raise_clean_exception()

**Comment by dbarrosop** [comment]
Solved via #299
